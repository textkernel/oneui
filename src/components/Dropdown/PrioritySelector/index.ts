export {
    PrioritySelector,
    Props as PrioritySelectorProps,
    PriorityItem,
    Priority,
} from './PrioritySelector';
