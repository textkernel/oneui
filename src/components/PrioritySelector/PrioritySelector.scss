@use 'sass:map';
@mixin common-button-style(
    $hoverColor: var(--color-background-neutral-subtlest-hover),
    $activeColor: var(--color-background-neutral-subtlest-pressed)
) {
    & {
        width: var(--space-400);
        height: var(--space-400);
        align-items: center;
        background-color: var(--transparent);
        cursor: pointer;
        display: flex;
        flex-shrink: 0;
        justify-content: center;
    }
    &:hover:not([disabled]) {
        background-color: $hoverColor;
    }
    &:active:not([disabled]) {
        background-color: $activeColor;
    }
    &[disabled] {
        cursor: not-allowed;
    }
}
@mixin selected-state(
    $selectedColor: var(--color-background-selected-subtlest-default),
    $hoverColor: var(--color-background-selected-subtlest-hover),
    $activeColor: var(--color-background-selected-subtlest-pressed)
) {
    & {
        background-color: $selectedColor;
    }
    &:hover:not([disabled]) {
        background-color: $hoverColor;
    }
    &:active:not([disabled]) {
        background-color: $activeColor;
    }
    &[disabled] {
        background-color: $selectedColor;
    }
}
$icon-status-colors: (
    mandatory: (
        normal: var(--color-icon-success-default),
        disabled: var(--color-icon-success-disabled),
    ),
    important: (
        normal: var(--color-icon-caution-default),
        disabled: var(--color-icon-caution-disabled),
    ),
    optional: (
        normal: var(--color-icon-subtle),
        disabled: var(--color-icon-disabled),
    ),
    exclude: (
        normal: var(--color-icon-critical-default),
        disabled: var(--color-icon-critical-disabled),
    ),
);
.PrioritySelector {
    width: 100%;
    @each $status, $colors in $icon-status-colors {
        &--#{$status} {
            fill: map.get($colors, normal);
            &[disabled] {
                fill: map.get($colors, disabled);
            }
        }
    }

    &__icon {
        outline: none;
        min-height: 20px;
        min-width: 20px;
        @each $status, $colors in $icon-status-colors {
            &--#{$status} {
                fill: map.get($colors, normal);
                &[disabled] {
                    fill: map.get($colors, disabled);
                }
            }
        }
    }
    &__priorityButton--isSelected,
    &__optionButton--isSelected {
        @include selected-state();
    }

    &__badgeListItem {
        display: flex;
        align-items: center;
        justify-content: flex-start;
        gap: var(--space-100);
    }
    &__badgeDropdownList {
        max-width: 400px;
        border-radius: var(--space-100);
        overflow: hidden;
        background-color: white;

        &--fixedWidth{
            width: 232px;
        }
    }

}




