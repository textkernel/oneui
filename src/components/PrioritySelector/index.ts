export {
    PrioritySelector,
    Props as PrioritySelectorProps,
    PriorityItemType,
    Priority,
} from './PrioritySelector';
