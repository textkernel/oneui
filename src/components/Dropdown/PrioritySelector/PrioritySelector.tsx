import * as React from 'react';
import Close from '@material-design-icons/svg/round/close.svg';
import KeyboardDoubleArrowUp from '@material-design-icons/svg/round/keyboard_double_arrow_up.svg';
import KeyboardArrowUp from '@material-design-icons/svg/round/keyboard_arrow_up.svg';
import KeyboardArrowDown from '@material-design-icons/svg/round/keyboard_arrow_down.svg';
import { bem } from '../../../utils/bem';
import { Dropdown } from '../Dropdown';
import { ListItem } from '../../List';
import { Text } from '../../Text';
import styles from './PrioritySelector.scss';
import { IconButton } from '../../Buttons';

const iconMap = {
    mandatory: KeyboardDoubleArrowUp,
    important: KeyboardArrowUp,
    optional: KeyboardArrowDown,
    exclude: Close,
};

export type Priority = 'mandatory' | 'important' | 'optional' | 'exclude';

export type PriorityItem = {
    priority: Priority;
    label: string;
    value?: string;
};

const defaultList: PriorityItem[] = [
    { priority: 'mandatory', label: 'Mandatory' },
    { priority: 'important', label: 'Important' },
    { priority: 'optional', label: 'Optional' },
    { priority: 'exclude', label: 'Exclude' },
];

export interface Props<>extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
    /** Currently selected priority item that indicates the importance of the component. */
    selectedItem: PriorityItem;
    /** Array of availible priority items. */
    list: PriorityItem[];
    /** Callback function triggered when a new priority is selected. */
    onChange: (newPriorityItem: PriorityItem) => void;
    /** Boolean indicating whether the whole badge should be disabled. */
    isDisabled?: boolean;
    /** Priority button label name for ARIA labelling */
    buttonLabel?: string;
}

const { block, elem } = bem('PrioritySelector', styles);
export function PrioritySelector({
    selectedItem,
    isDisabled = false,
    onChange,
    buttonLabel,
    list = defaultList,
    ...rest
}: Props) {
    const badgeRef = React.useRef<HTMLDivElement | null>(null);

    const hasPriorityList = list && list.length > 0;

    const renderPriorityIcon = (priorityType?: Priority, disabled: boolean = false) => {
        if (!priorityType) {
            return null;
        }

        const IconComponent = iconMap[priorityType];

        return IconComponent ? (
            <IconComponent
                {...elem('icon', { [priorityType]: true })}
                disabled={disabled}
                viewBox="0 0 24 24"
            />
        ) : null;
    };

    return (
        <div {...block({ ...rest })} ref={badgeRef}>
            {hasPriorityList && (
                <Dropdown<PriorityItem>
                    button={
                        <IconButton
                            {...elem('icon', { [selectedItem.priority]: true })}
                            variant="ghost"
                            aria-label={buttonLabel}
                            disabled={isDisabled}
                            type="button"
                        >
                            {renderPriorityIcon(selectedItem.priority, isDisabled)}
                        </IconButton>
                    }
                    onChange={(newPriorityItem) => onChange(newPriorityItem)}
                    placement="bottom-start"
                    listClassName={styles.badgeDropdownList}
                    refElement={badgeRef}
                >
                    {list.map((item) => (
                        <ListItem className={styles.badgeListItem} key={item.priority} value={item}>
                            {renderPriorityIcon(item.priority)}
                            <Text inline size="small">
                                {item.label}
                            </Text>
                        </ListItem>
                    ))}
                </Dropdown>
            )}
        </div>
    );
}

PrioritySelector.displayName = 'PrioritySelector';
