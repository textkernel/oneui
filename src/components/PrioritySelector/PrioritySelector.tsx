import * as React from 'react';
import Close from '@material-design-icons/svg/round/close.svg';
import KeyboardDoubleArrowUp from '@material-design-icons/svg/round/keyboard_double_arrow_up.svg';
import KeyboardArrowUp from '@material-design-icons/svg/round/keyboard_arrow_up.svg';
import KeyboardArrowDown from '@material-design-icons/svg/round/keyboard_arrow_down.svg';
import { DropdownMenuItem, Root, Portal } from '@radix-ui/react-dropdown-menu';
import { Size } from '@textkernel/oneui';
import { DropdownTrigger } from '../Dropdown/DropdownTrigger';
import { DropdownContent } from '../Dropdown/DropdownContent';
import { bem } from '../../utils/bem';
import { Text } from '../Text';
import { IconButton } from '../Buttons';
import itemStyles from '../Dropdown/Items/Item.scss';
import styles from './PrioritySelector.scss';

const iconMap = {
    mandatory: KeyboardDoubleArrowUp,
    important: KeyboardArrowUp,
    optional: KeyboardArrowDown,
    exclude: Close,
};

export type Priority = 'mandatory' | 'important' | 'optional' | 'exclude';

export type PriorityItemType<PriorityItemValue> = {
    /** priority types: mandatory, important, optional or exclude. Makes correct icon show up */
    priority: Priority;
    /** text that should accompany the icon in the dropdown */
    label: string;
    /** optional: used in case of application-specific values / custom priorities */
    value?: PriorityItemValue;
};

export interface Props<PriorityItemValue>
    extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
    /** Currently selected priority item that indicates the importance of the component. */
    selectedItem: PriorityItemType<PriorityItemValue>;
    /** Array of availible priority items. */
    list: PriorityItemType<PriorityItemValue>[];
    /** Callback function triggered when a new priority is selected. */
    onChange: (newPriorityItem: PriorityItemType<PriorityItemValue>) => void;
    /** Boolean indicating whether the whole badge should be disabled. */
    isDisabled?: boolean;
    /** Priority button label name for ARIA labelling */
    buttonLabel?: string;
    /** ref to possible parent div, if it is a badge */
    parentRef?: React.RefObject<HTMLElement | null> | React.ForwardedRef<HTMLDivElement>;
    /** ref to iconButton, used for keyboard navigation */
    buttonRef?: React.RefObject<HTMLElement>;
    /** the size of the trigger button */
    size?: Size;
}

const { block, elem } = bem('PrioritySelector', styles);
const itemStylesBem = bem('DropdownItem', itemStyles);
export function PrioritySelector<PriorityItemValue>({
    onChange,
    selectedItem,
    isDisabled = false,
    buttonLabel,
    list,
    parentRef,
    buttonRef,
    size,
    ...rest
}: Props<PriorityItemValue>) {
    const renderPriorityIcon = (
        priorityType?: Priority,
        disabled: boolean = false,
        inList?: boolean
    ) => {
        if (!priorityType) {
            return null;
        }
        const IconComponent = iconMap[priorityType];
        return IconComponent ? (
            <IconComponent
                {...elem('icon', { [priorityType]: true, inList })}
                disabled={disabled}
                viewBox="0 0 24 24"
            />
        ) : null;
    };

    const priorityOrder: Record<Priority, number> = {
        mandatory: 1,
        important: 2,
        optional: 3,
        exclude: 4,
    };

    // Sort the list based on the priority order
    const sortedList = list
        .slice()
        .sort((a, b) => priorityOrder[a.priority] - priorityOrder[b.priority]);

    return (
        <Root {...block({ ...rest })} refElement={parentRef}>
            <DropdownTrigger>
                <IconButton
                    {...elem('icon', { [selectedItem.priority]: true })}
                    variant="ghost"
                    aria-label={buttonLabel || selectedItem.label}
                    disabled={isDisabled}
                    type="button"
                    ref={buttonRef}
                    size={size}
                >
                    {renderPriorityIcon(selectedItem.priority, isDisabled)}
                </IconButton>
            </DropdownTrigger>
            <Portal>
                <DropdownContent
                    {...elem('badgeDropdownList', { fixedWidth: !parentRef })}
                    sideOffset={6}
                    refElement={parentRef}
                >
                    {sortedList.map((item) => (
                        <DropdownMenuItem
                            key={item.priority}
                            onSelect={(e) => {
                                e.stopPropagation();
                                onChange(item);
                            }}
                            role="option"
                            aria-selected={selectedItem.priority === item.priority}
                            tabIndex={isDisabled ? -1 : 0}
                            {...rest}
                            {...itemStylesBem.block({
                                isSelected: selectedItem === item,
                                disabled: isDisabled,
                                ...rest,
                            })}
                        >
                            <div {...elem('badgeListItem')}>
                                {renderPriorityIcon(item.priority, false, true)}
                                <Text inline size="small">
                                    {item.label}
                                </Text>
                            </div>
                        </DropdownMenuItem>
                    ))}
                </DropdownContent>
            </Portal>
        </Root>
    );
}

PrioritySelector.displayName = 'PrioritySelector';
