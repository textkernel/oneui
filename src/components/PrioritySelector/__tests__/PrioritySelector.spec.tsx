import * as React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { userEvent } from '@testing-library/user-event';
import { PriorityItemType, PrioritySelector } from '../PrioritySelector';

const priorityList: PriorityItemType<string>[] = [
    { priority: 'mandatory', label: 'Mandatory', value: 'required' },
    { priority: 'important', label: 'Important', value: 'strongly_favored' },
    { priority: 'optional', label: 'Optional', value: 'favored' },
    { priority: 'exclude', label: 'Exclude', value: 'banned' },
];

const priorityListRandom: PriorityItemType<string>[] = [
    { priority: 'optional', label: 'Optional', value: 'favored' },
    { priority: 'mandatory', label: 'Mandatory', value: 'required' },
    { priority: 'exclude', label: 'Exclude', value: 'banned' },
    { priority: 'important', label: 'Important', value: 'strongly_favored' },
];

describe('PrioritySelector', () => {
    let view: RenderResult;
    const onSelectMock = jest.fn();

    beforeEach(() => {
        view = render(
            <PrioritySelector
                selectedItem={priorityList[0]}
                list={priorityList}
                onChange={onSelectMock}
            />
        );
    });
    afterEach(() => {
        jest.resetAllMocks();
    });

    it('renders correctly with all props provided', () => {
        expect(view.container).toMatchSnapshot();
        expect(screen.getByRole('button', { name: 'Mandatory' })).toBeInTheDocument();
    });

    it('should render correctly opened', async () => {
        const user = userEvent.setup();
        await user.click(screen.getByRole('button', { name: 'Mandatory' }));
        expect(view.baseElement).toMatchSnapshot();
        expect(screen.getAllByRole('option')).toHaveLength(4);
    });

    it('should call onSelect correctly', async () => {
        const user = userEvent.setup();
        expect(onSelectMock).toHaveBeenCalledTimes(0);

        await user.click(screen.getByRole('button', { name: 'Mandatory' }));
        await user.click(screen.getByText('Important'));

        expect(onSelectMock).toHaveBeenCalledTimes(1);
        expect(onSelectMock).toHaveBeenCalledWith(priorityList[1]);
    });

    it('should close dropdown when item is chosen', async () => {
        const user = userEvent.setup();
        await user.click(screen.getByRole('button', { name: 'Mandatory' }));

        expect(screen.getByRole('menu')).toHaveAttribute('data-state', 'open');
        await user.click(screen.getAllByRole('option')[1]);

        expect(screen.getByRole('button')).toHaveAttribute('data-state', 'closed');
    });

    it('should render the priorities in the correct order', async () => {
        const user = userEvent.setup();
        await user.click(screen.getByRole('button', { name: 'Mandatory' }));

        expect(screen.getAllByTestId('default-icon')[1]).toHaveClass(
            'PrioritySelector__icon--mandatory'
        );
        expect(screen.getAllByTestId('default-icon')[2]).toHaveClass(
            'PrioritySelector__icon--important'
        );
        expect(screen.getAllByTestId('default-icon')[3]).toHaveClass(
            'PrioritySelector__icon--optional'
        );
        expect(screen.getAllByTestId('default-icon')[4]).toHaveClass(
            'PrioritySelector__icon--exclude'
        );

        view.rerender(
            <PrioritySelector
                selectedItem={priorityListRandom[0]}
                list={priorityListRandom}
                onChange={onSelectMock}
            />
        );

        expect(screen.getAllByTestId('default-icon')[1]).toHaveClass(
            'PrioritySelector__icon--mandatory'
        );
        expect(screen.getAllByTestId('default-icon')[2]).toHaveClass(
            'PrioritySelector__icon--important'
        );
        expect(screen.getAllByTestId('default-icon')[3]).toHaveClass(
            'PrioritySelector__icon--optional'
        );
        expect(screen.getAllByTestId('default-icon')[4]).toHaveClass(
            'PrioritySelector__icon--exclude'
        );
    });
});
