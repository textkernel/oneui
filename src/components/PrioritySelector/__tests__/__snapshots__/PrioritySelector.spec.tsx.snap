// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrioritySelector renders correctly with all props provided 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    aria-label="Mandatory"
    class="Button Button--variant_ghost Button--context_secondary IconButton--size_medium PrioritySelector__icon PrioritySelector__icon--mandatory test-class"
    data-state="closed"
    id="radix-:r0:"
    type="button"
  >
    <svg
      class="IconButton__icon--size_medium"
      data-testid="default-icon"
      viewBox="0 0 24 24"
    />
  </button>
</div>
`;

exports[`PrioritySelector should render correctly opened 1`] = `
<body
  data-scroll-locked="1"
  style="pointer-events: none;"
>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <button
      aria-controls="radix-:r3:"
      aria-expanded="true"
      aria-haspopup="menu"
      aria-label="Mandatory"
      class="Button Button--variant_ghost Button--context_secondary IconButton--size_medium PrioritySelector__icon PrioritySelector__icon--mandatory test-class"
      data-state="open"
      id="radix-:r2:"
      type="button"
    >
      <svg
        class="IconButton__icon--size_medium"
        data-testid="default-icon"
        viewBox="0 0 24 24"
      />
    </button>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    dir="ltr"
    style="position: fixed; left: 0px; top: 0px; transform: translate(0px, 6px); min-width: max-content; --radix-popper-available-width: 0px; --radix-popper-available-height: -6px; --radix-popper-anchor-width: 0px; --radix-popper-anchor-height: 0px; --radix-popper-transform-origin: 0% 0px;"
  >
    <div
      aria-labelledby="radix-:r2:"
      aria-orientation="vertical"
      class="DropdownContent PrioritySelector__badgeDropdownList PrioritySelector__badgeDropdownList--fixedWidth"
      data-align="start"
      data-orientation="vertical"
      data-radix-menu-content=""
      data-side="bottom"
      data-state="open"
      dir="ltr"
      id="radix-:r3:"
      role="menu"
      style="outline: none; --radix-dropdown-menu-content-transform-origin: var(--radix-popper-transform-origin); --radix-dropdown-menu-content-available-width: var(--radix-popper-available-width); --radix-dropdown-menu-content-available-height: var(--radix-popper-available-height); --radix-dropdown-menu-trigger-width: var(--radix-popper-anchor-width); --radix-dropdown-menu-trigger-height: var(--radix-popper-anchor-height); pointer-events: auto;"
      tabindex="-1"
    >
      <div
        aria-selected="true"
        class="DropdownItem DropdownItem--isSelected"
        data-orientation="vertical"
        data-radix-collection-item=""
        role="option"
        tabindex="0"
      >
        <div
          class="PrioritySelector__badgeListItem"
        >
          <svg
            class="PrioritySelector__icon PrioritySelector__icon--mandatory PrioritySelector__icon--inList"
            data-testid="default-icon"
            viewBox="0 0 24 24"
          />
          <span
            class="Text--size_small"
          >
            Mandatory
          </span>
        </div>
      </div>
      <div
        aria-selected="false"
        class="DropdownItem"
        data-orientation="vertical"
        data-radix-collection-item=""
        role="option"
        tabindex="0"
      >
        <div
          class="PrioritySelector__badgeListItem"
        >
          <svg
            class="PrioritySelector__icon PrioritySelector__icon--important PrioritySelector__icon--inList"
            data-testid="default-icon"
            viewBox="0 0 24 24"
          />
          <span
            class="Text--size_small"
          >
            Important
          </span>
        </div>
      </div>
      <div
        aria-selected="false"
        class="DropdownItem"
        data-orientation="vertical"
        data-radix-collection-item=""
        role="option"
        tabindex="0"
      >
        <div
          class="PrioritySelector__badgeListItem"
        >
          <svg
            class="PrioritySelector__icon PrioritySelector__icon--optional PrioritySelector__icon--inList"
            data-testid="default-icon"
            viewBox="0 0 24 24"
          />
          <span
            class="Text--size_small"
          >
            Optional
          </span>
        </div>
      </div>
      <div
        aria-selected="false"
        class="DropdownItem"
        data-orientation="vertical"
        data-radix-collection-item=""
        role="option"
        tabindex="0"
      >
        <div
          class="PrioritySelector__badgeListItem"
        >
          <svg
            class="PrioritySelector__icon PrioritySelector__icon--exclude PrioritySelector__icon--inList"
            data-testid="default-icon"
            viewBox="0 0 24 24"
          />
          <span
            class="Text--size_small"
          >
            Exclude
          </span>
        </div>
      </div>
    </div>
  </div>
  <span
    aria-hidden="true"
    data-aria-hidden="true"
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;
